[dtensor] refactor sharding cost model to count for latency

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119897

This PR refactors the shardeing cost model, to do a more accurate
estimation of redistribute cost, including both collective latency and
communciation time.

The previous cost model does not recale the latency and communciation
time, therefore the latency factor is too small to be counted, and in
the case of small tensors, multiple collectives is preferred than a
single collective, which is wrong.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225

wanchaol committed:
[dtensor] refactor sharding cost model to count for latency

This PR refactors the shardeing cost model, to do a more accurate
estimation of redistribute cost, including both collective latency and
communciation time.

The previous cost model does not recale the latency and communciation
time, therefore the latency factor is too small to be counted, and in
the case of small tensors, multiple collectives is preferred than a
single collective, which is wrong.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] refactor sharding cost model to count for latency"


This PR refactors the shardeing cost model, to do a more accurate
estimation of redistribute cost, including both collective latency and
communciation time.

The previous cost model does not recale the latency and communciation
time, therefore the latency factor is too small to be counted, and in
the case of small tensors, multiple collectives is preferred than a
single collective, which is wrong.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

tianyu-l commented on torch/distributed/_tensor/_collective_utils.py:
```diff
@@ -167,65 +169,85 @@ def spec_to_bytes(spec: "placement_types.DTensorSpec") -> int:
+def allreduce_cost(bytes_gb: float, mesh_topo: MeshTopoInfo, mesh_dim: int) -> float:
+    num_devices_on_mesh_dim = mesh_topo.mesh_dim_devices[mesh_dim]
+    mesh_dim_bandwidth = mesh_topo.mesh_dim_bandwidth[mesh_dim]
     # allreduce have 2x comm bytes compare to allgather/reduce_scatter
```
pls change this comment as more subtle difference (2x vs. 2x - 1) is captured now

tianyu-l reviewed (approved):
Logic seems mostly correct (had some inline comments), although I haven't verified the numbers. Let me know if you'd like someone to double check the numbers.

wanchaol commented on torch/distributed/_tensor/_collective_utils.py:
```diff
@@ -167,65 +169,85 @@ def spec_to_bytes(spec: "placement_types.DTensorSpec") -> int:
+                # This number assumes latest GPU arch, i.e. Ampere or Hopper
+                # TODO: see if we need to tweak this or offer a way for user
+                # to specify the bandwidths/latency
+                mesh_dim_bandwidth[mesh_dim] *= 0.22
```
err, that's true, this should be inter host comm bandwidth

wanchaol committed:
Update on "[dtensor] refactor sharding cost model to count for latency"


This PR refactors the shardeing cost model, to do a more accurate
estimation of redistribute cost, including both collective latency and
communciation time.

The previous cost model does not recale the latency and communciation
time, therefore the latency factor is too small to be counted, and in
the case of small tensors, multiple collectives is preferred than a
single collective, which is wrong.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

awgu commented on torch/distributed/_tensor/_collective_utils.py:
```diff
@@ -167,65 +169,85 @@ def spec_to_bytes(spec: "placement_types.DTensorSpec") -> int:
+        # 3. we assume collectives are all ring base algo for now
+        num_devices_per_host = _mesh_resources.num_devices_per_host(mesh.device_type)
+        # the base bw number (intra-node), GB/s
+        base_bw = 87.7
```
Where did we get this intra-node bandwidth number?

wanchaol replied:
mostly from here https://github.com/pytorch/pytorch/blob/main/torch/_inductor/comm_analysis.py#L156

awgu replied:
Since these numbers are per channel and NCCL uses multiple channels, would this base bandwidth number be low? E.g. normally intra-node bandwidth on Ampere is ~300 GB/s I thought.

wanchaol replied:
Yeah thanks for catching this! I'm trying to figure out what exactly is the per-GPU bandwidth to other GPUs, in the case of NVLink. 300 GB/s means uni-directional p2p bandwidth, is this counting the channels already?

There's some bus bandwidth calculation where it counts for # channels = 2, and have some bandwith refinements, if we take this formula down for intra-node then it's around 2 * 300 /3 = 200 https://github.com/pytorch/pytorch/blob/main/torch/_inductor/comm_analysis.py#L223-L227 I am not really sure it should be 200 here, as we are really caculating the bus bandwidth from an algorithm bandwidth. So for nvlink if the algo bwdwidth should be 300 then the base_bw here should be 300 instead I guess?

wanchaol replied:
🤔 maybe the comm analysis pass bandwidth estimation is kinda wrong too, as it would almost always take the smaller one, `llMaxBw` which is 87.6
